crates/formats/src/param.rs

```diff
@@ -0,0 +1,594 @@
+    use zerocopy::{ByteOrder, FromBytes, Unaligned, U16};
+
+    pub trait StaticBO: 'static + ByteOrder {}
+    impl<BO: 'static + ByteOrder> StaticBO for BO {}
+
+    pub trait UnalignedInto<T>: Into<T> + Copy + Unaligned + FromBytes {}
+    impl<T, U: Into<T> + Copy + Unaligned + FromBytes> UnalignedInto<T> for U {}
+
+    /// A valid Unicode string slice which may or may not be UTF-8 encoded.
+    pub trait GenericStr: ToString {
+        /// Returns the length of the string slice in bytes.
+        fn len_bytes(&self) -> usize;
+
+        /// Attempt to create a string slice from a null-terminated sequence of bytes.
+        fn read_cstr(bytes: &[u8]) -> Option<&'_ Self>;
+
+        /// Convert to a Rust utf-8 string slice, possibly performing a conversion.
+        fn to_rust_str(&self) -> Cow<'_, str>;
+    }
+
+    impl GenericStr for str {
+        fn len_bytes(&self) -> usize {
+            self.as_bytes().len()
+        }
+
+        fn read_cstr(bytes: &[u8]) -> Option<&'_ Self> {
+            let cstr = CStr::from_bytes_until_nul(bytes).ok()?;
+            cstr.to_str().ok()
+        }
+
+        fn to_rust_str(&self) -> Cow<'_, str> {
+            Cow::Borrowed(self)
+        }
+    }
+
+    impl<BO: StaticBO> GenericStr for WStr<BO> {
+        fn len_bytes(&self) -> usize {
+            self.as_bytes().len()
+        }
+
+        fn read_cstr(bytes: &[u8]) -> Option<&'_ Self> {
+            let nt_pos = bytes.chunks_exact(2).position(|b| b == &[0, 0])?;
+            WStr::from_utf16(&bytes[..2 * nt_pos]).ok()
+        }
+
+        fn to_rust_str(&self) -> Cow<'_, str> {
+            Cow::Owned(self.to_string())
+        }
+    }
+
+    pub trait CharType<BO: StaticBO> {
+        type Unit: UnalignedInto<u32>;
+        type Str: GenericStr + ?Sized;
+    }
+
+    pub trait OffsetType<BO: StaticBO> {
+        type T: UnalignedInto<u64>;
+        type Pad: Unaligned + FromBytes;
+    }
+
+    /// Marker type used to represent a param file with 32-bit offsets.
+    pub struct Offset32 {}
+    impl<BO: StaticBO> OffsetType<BO> for Offset32 {
+        type T = zerocopy::U32<BO>;
+        type Pad = [u8; 0];
+    }
+
+    /// Marker type used to represent a param file with 64-bit offsets.
+    pub struct Offset64 {}
+    impl<BO: StaticBO> OffsetType<BO> for Offset64 {
+        type T = zerocopy::U64<BO>;
+        type Pad = [u8; 4];
+    }
+
+    /// Marker type used to represent a param file with UTF-8 strings.
+    pub struct Char {}
+    impl<BO: StaticBO> CharType<BO> for Char {
+        type Unit = u8;
+        type Str = str;
+    }
+
+    /// Marker type used to represent a param file with UTF-16 strings.
+    pub struct WChar {}
+    impl<BO: StaticBO> CharType<BO> for WChar {
+        type Unit = U16<BO>;
+        type Str = WStr<BO>;
+    }
+
+    /// Trait containing associated types that vary according to the endianness,
+    /// offset size and string encoding of a param file.
+    pub trait ParamTraits {
```

Comment:
Not a fan of `ParamTraits`, how about something like `ParamLayout` (or perhaps something a little less loaded that  could be confused with row layout)?

Reply:
`ParamLayout` makes sense to me. While it could be confused with something like a paramdef, I do prefer it to `ParamTraits` and can't really think of anything better. The reason for `Traits` was C++ terminology but it doesn't make much sense here

Reply:
Decided to go with `ParamFileLayout` to reduce ambiguity with paramdefs.